### ISSUE TITLE
Configuration files randomly grow in size up to several gigabytes for no reason

### ISSUE THREAD

**Issue by BoomEaro** [bergerhealer/BKCommonLib#121]
##### Spigot version: paper 1.16.5 build 573

<hr>

#### Problem or bug:
Configuration files specifically for BKCommonLib and plugins using BKCommonLib to create their yml configurations can sometimes grow abnormally up to several gigabytes for no reason.
https://imgur.com/a/ryWXW3b
https://imgur.com/a/BNhRIA5
(The screenshot shows the files of the Light Cleaner plugin that uses BKCommonLib)

#### Expected behaviour:
Files should not endlessly grow in size for no reason

#### Steps to reproduce:
To be honest, I don't understand how exactly to repeat this problem.
The first time this came up was when I killed the server because it was frozen for some other reason. Then I started to notice that the occupied disk space suddenly increased.
The server ran on 11 java and windows 8.1. (could this possibly be related to the operating system?)

The most important thing is that this problem is only with BKCommonLib and plugins that use it, other plugins using yml from bukkit work fine.

**Comment by bergerkiller** [comment]
In config.yml of bkcommonlib is an option to write configs sync. Does turning this on fix this problem for you?

The random trailing null bytes is a JVM issue with their async io api, it appears to be machine-specific and I havent been able to reproduce it myself. A select few lucky people run into this.

**Comment by BoomEaro** [comment]
Sorry for not looking at config.yml, I was so busy with the problem and experimenting with deleting these configuration files a bunch of times that I simply did not read the description of each option in config.yml.

Set forceSynchronousSaving: true and seems to be working correctly, thanks!

**Comment by bergerkiller** [comment]
Can leave this issue ticket open, its still an issue, just one I haven't found a solution for yet because the issue occurs in the JVM, or maybe even OS filesystem driver, or maybe even a virtual hypervisor thing. If you have a clue what might be causing it, any information is helpful :)

**Comment by BoomEaro** [comment]
I have Avast antivirus installed, but even though I disabled various protection components (file or behavior analysis), this did not solve the problem. But, perhaps, the very presence of an antivirus in the system can affect this.

It seems that the java version does not really matter, I used to have 8 java and then I also had such a problem, but then I didn't really attach much importance to it.

As I already said, the server was launched on Windows 8.1, although in real conditions the servers are usually on linux.
Perhaps this is just a problem with the old OS.

**Comment by bergerkiller** [comment]
Well, if in the future more people have  this same issue that also turn out to be using windows 8.1 or avast then we can at least lay some correlation. I believe a previous person that had this problem also used Windows.

This can still be a problem with how I wrote the async i/o code. Java really doesn't make asynchronous i/o easy and its full of "caution: here be dragons" as I ran into strange jvm hitches.

Might be worth adding a third mode that just writes the file on another thread rather than using [async nio](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/AsynchronousChannel.html)
